modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -21,116 +21,194 @@ import net.corda.core.node.services.vault.QueryCriteria
 import net.corda.core.serialization.SingletonSerializeAsToken
 import net.corda.core.utilities.contextLogger
 import rx.Observable
+import rx.subjects.PublishSubject
 import java.time.Duration
-import java.util.concurrent.ConcurrentHashMap
-import java.util.concurrent.ConcurrentMap
-import java.util.concurrent.Executors
-import java.util.concurrent.ScheduledExecutorService
-import java.util.concurrent.TimeUnit
+import java.util.concurrent.*
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+val UPDATER: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
```

Comment:
Here we use the same (single) thread for updating the caches as well as unlocking? could we end up with some blocking here if the thread is constantly busy? Maybe we can parallelise?

Reply:
so fingers crossed, adding a token to a bucket is MUCH MUCH MUCH faster than producing or consuming tokens, as it's just updating maps - whilst production/consumption involves flows etc etc. 

Initial testing is that 1 updater thread can keep up with >100 flows producing tokens without breaking a sweat.

Reply:
ok awesome - thanks